app/src/main/java/com/github/drawyourpath/bootcamp/authentication/FirebaseAuth.kt

```diff
@@ -0,0 +1,236 @@
+            currCallback!!(user, error)
+            currCallback = null
+        }
+    }
+
+    override fun getUser(): User? {
+        return convertUser(FirebaseAuth.getInstance().currentUser)
+    }
+
+    override fun loginWithGoogle(activity: Activity, callback: AuthCallback) {
+        if (!setCurrCallback(callback)) {
+            return
+        }
+
+        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
+            .requestIdToken(activity.getString(R.string.server_client_id))
+            .requestEmail()
+            .build()
+        val client = GoogleSignIn.getClient(activity, gso)
+
+        client.signInIntent.also {
+            activity.startActivityForResult(it, REQ_GSI)
+        }
+    }
+
+    override fun loginWithEmail(email: String, password: String, callback: AuthCallback) {
+        TODO("Not yet implemented")
+    }
+
+    override fun loginAnonymously(callback: AuthCallback) {
+        TODO("Not yet implemented")
+    }
+
+    override fun registerWithEmail(email: String, password: String, callback: AuthCallback) {
+        auth
+            .createUserWithEmailAndPassword(email, password)
+            .addOnSuccessListener {res ->
+                callback(convertUser(res.user!!), null)
+            }
+            .addOnFailureListener { exc ->
+                callback(null, exc)
+            }
+    }
+
+    override fun registerWithGoogle(activity: Activity, callback: AuthCallback) {
+        loginWithGoogle(activity, callback)
+    }
+
+    private var authStateListener: FirebaseAuth.AuthStateListener? = null;
+
+    override fun onAuthStateChanged(callback: AuthCallback) {
+        if (authStateListener != null)
+        {
+            clearListener()
+        }
+
+        authStateListener = FirebaseAuth.AuthStateListener {
+            callback(getUser(), null)
+        }
+
+        auth.addAuthStateListener(authStateListener!!)
+    }
+
+    override fun clearListener() {
+        if (authStateListener != null) {
+            auth.removeAuthStateListener(authStateListener!!)
+            authStateListener = null;
+        }
+    }
+
+    override fun launchOneTapGoogleSignIn(activity: Activity, callback: AuthCallback) {
+        if (!setCurrCallback(callback)) {
+            return
+        }
+
+        oneTapClient.beginSignIn(signInRequest)
+            .addOnSuccessListener(activity) { result ->
+                try {
+                    activity.startIntentSenderForResult(
+                        result.pendingIntent.intentSender, REQ_ONE_TAP,
+                        null, 0, 0, 0, null)
+                } catch (e: IntentSender.SendIntentException) {
+                    Log.e("GSI", "Couldn't start One Tap UI: ${e.localizedMessage}")
+                }
+            }
+            .addOnFailureListener(activity) { e ->
+                Log.d("GSI", e.localizedMessage!!)
+            }
+    }
+
+    override fun signOut() {
+        auth.signOut()
+    }
```

Comment:
Maybe set the currCallback to null?

Reply:
`currCallback` is automatically cleared after the sign in procedure completes. If we set it to null in `signOut`, it will cause a sign in to not call its callback or just do nothing as it is already `null`.